lib/SPIRV/SPIRVUtil.cpp

```diff
@@ -701,61 +882,37 @@ void getParameterTypes(Function *F, SmallVectorImpl<StructType *> &ArgTys) {
-  if (Demangler.partialDemangle(MangledName.c_str()))
+  const StringRef MangledName(F->getName());
+  ManglingParser<DefaultAllocator> Demangler(MangledName.begin(),
+                                             MangledName.end());
+  // We expect to see only function name encodings here. If it's not a function
+  // name encoding, bail out.
+  auto *RootNode = dyn_cast_or_null<FunctionEncoding>(Demangler.parse());
+  if (!RootNode)
     return;
-  char *Buf = nullptr;
-  size_t BufLen = 0;
-  Buf = Demangler.getFunctionParameters(Buf, &BufLen);
-  StringRef Args(Buf, BufLen);
-  // Strip parentheses from the result.
-  Args = Args.slice(1, Args.size() - 2);
-
-  if (Args.find_first_of("<(") == StringRef::npos) {
-    // Go through the function arguments using type names where possible.
-    SmallVector<StringRef, 8> ArgParams;
-    Args.split(ArgParams, ", ");
-
-    // Sanity check that the name mangling matches up to the expected number of
-    // arguments.
-    if (ArgParams.size() > (size_t)(ArgTys.end() - ArgIter)) {
-      LLVM_DEBUG(dbgs() << "[getParameterTypes] function " << F->getName()
-                        << " appears to have " << ArgParams.size()
-                        << " arguments but has " << (ArgTys.end() - ArgIter)
-                        << "\n");
-      free(Buf);
-      return;
-    }
 
-    for (StringRef Arg : ArgParams) {
-      StructType *Pointee = nullptr;
-      if (Arg.endswith("*") && !Arg.endswith("**")) {
-        // Strip off address space and other qualifiers.
-        StringRef MangledStructName = Arg.split(' ').first;
-
-        if (MangledStructName.consume_front("__spirv_")) {
-          // This is a pointer to a SPIR-V OpType* opaque struct. In general,
-          // convert __spirv_<Type>[__Suffix] to %spirv.Type[._Suffix]
-          auto NameSuffixPair = MangledStructName.split('_');
-          std::string StructName = "spirv.";
-          StructName += NameSuffixPair.first;
-          if (!NameSuffixPair.second.empty()) {
-            StructName += ".";
-            StructName += NameSuffixPair.second;
-          }
-          Pointee = getOrCreateOpaqueStructType(M, StructName);
-        } else if (MangledStructName.startswith("opencl.")) {
-          Pointee = getOrCreateOpaqueStructType(M, MangledStructName);
-        }
-      } else if (!Arg.contains(' ') && Arg.startswith("ocl_")) {
-        std::string StructName = demangleBuiltinOpenCLTypeName(Arg);
-        Pointee = getOrCreateOpaqueStructType(M, StructName);
-      }
-      *ArgIter++ = Pointee;
+  // Sanity check that the name mangling matches up to the expected number of
+  // arguments.
+  if (RootNode->getParams().size() > (size_t)(ArgTys.end() - ArgIter)) {
+    LLVM_DEBUG(dbgs() << "[getParameterTypes] function " << MangledName
+                      << " appears to have " << RootNode->getParams().size()
+                      << " arguments but has " << (ArgTys.end() - ArgIter)
+                      << "\n");
```

Comment:
Here we should probably notify llvm-spirv user and not llvm-spirv developer, that something went very wrong. Should it be `unreachable` or any other error reporting mechanism instead of LLVM_DEBUG?

Reply:
This function will, in certain cases, be called on user-defined functions (as opposed to just SPIR-V/OpenCL builtins), so I'm reluctant to make it be a hard fail, in case someone decides to name their OpenCL kernel function something which is a valid Itanium name mangling that doesn't match their function signature. While reporting to the llvm-spirv user in lieu of the developer would be useful, I'm not sure if there's an appropriate mechanism for that.

Reply:
Got it, thanks!

>> While reporting to the llvm-spirv user in lieu of the developer would be useful, I'm not sure if there's an appropriate mechanism for that.

Yeah, I remember having some internal tracker for that :)